src/Listener/HealthCheckListener.php

```diff
@@ -0,0 +1,71 @@
+<?php
+
+namespace PrestaShop\Module\HealthCheck\Listener;
+
+use PrestaShop\Module\HealthCheck\Model\HealthCheckConfig;
+use PrestaShop\Module\HealthCheck\Repository\HealthCheckConfigRepository;
+use PrestaShop\PrestaShop\Adapter\LegacyContext;
+use Symfony\Component\HttpFoundation\IpUtils;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+
+class HealthCheckListener
+{
+    /**
+     * @var LegacyContext
+     */
+    private $context;
+    /**
+     * @var HealthCheckConfigRepository
+     */
+    private $healthCheckConfigRepository;
+
+    public function __construct(LegacyContext $context, HealthCheckConfigRepository $healthCheckConfigRepository)
+    {
+        $this->context = $context;
+        $this->healthCheckConfigRepository = $healthCheckConfigRepository;
+    }
+
+    /**
+     * Check if the user is allowed to access the healthcheck route.
+     *
+     * @return bool|null or redirect
+     */
+    public function onKernelRequest(GetResponseEvent $event)
+    {
+        $request = $event->getRequest();
+        if ('_admin_healthcheck' !== $request->get('_route')) {
+            return null;
+        }
+
+        $configId = $this->healthCheckConfigRepository->getLastId();
+        $healthCheckConfig = new HealthCheckConfig($configId);
+
+        if (
+            !empty($healthCheckConfig->ips) &&
+            !empty($healthCheckConfig->token) &&
+            true === IpUtils::checkIp(
+                $request->getClientIp(),
+                array_map('trim', explode(',', $healthCheckConfig->ips))
+            ) &&
+            $request->headers->has('X-HEALTHCHECK-TOKEN') &&
+            0 === strcmp($healthCheckConfig->token, $request->headers->get('X-HEALTHCHECK-TOKEN'))
+        ) {
+            return true;
+        }
+
+        //employee not logged in
+        $event->stopPropagation();
+
+        //if http request - add 403 error
+        $request = Request::createFromGlobals();
+        if ($request->isXmlHttpRequest()) {
+            header('HTTP/1.1 403 Forbidden');
+            exit();
```

Comment:
Never use exist like that. Use Response object instead

Reply:
I did the same with https://github.com/PrestaShop/PrestaShop/blob/develop/src/Adapter/Security/Admin.php#L102
I changed it to RedirectResponse ;)

Reply:
Security/Admin is an old file, not using SymfonyKernel events :( (AFAIK or using it badly)